### ISSUE TITLE
gunzip middleware is unable to decode SDCH content-encoding

### ISSUE THREAD

**Issue by shaunlebron** [joeferner/node-http-mitm-proxy#80]
I am able to run the [example](https://github.com/joeferner/node-http-mitm-proxy#example), but I'm not sure how to test it out in my browser.  How do I get my browser to use this proxy server?

Trying to set Chrome's proxy server redirected me to do this in my network settings. Frankly, I've never created/used a proxy.
![screen shot 2016-05-05 at 10 04 54 pm](https://cloud.githubusercontent.com/assets/116838/15063189/8495e444-130d-11e6-939a-82895c660710.png)

**Comment by felicienfrancois** [comment]
To setup the proxy in your browser only, the easiest way is to use an extension.
For example SwitchyOmega for Google Chrome is really easy to use and allow you to quickly switch between proxy configurations.
The settings are:
- Kind of proxy: HTTP
- Host: localhost (or 127.0.0.1)
- Port: the port you passed in parameters (8081 in the example)

To setup the proxy in Mac OS X (apply to all http requests including ones from the OS, browsers or third party apps), you've found the right setting (Wi-fi > Proxies > Web Proxy (HTTP) > localhost:8081).

You will also need to either disable web security (using chrome flags `--disable-web-security --ignore-certificate-errors`) or by trusting the certificate autority (recommended).
This is because node-http-mitm-proxy is a [mitm](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) proxy. it forge certificates for each browsed domains to be able to intercept the encrypted traffic. Thoose certificates are forged with a self generated Certificate Authority which is obviously not trusted by OS and browsers.

By default, you will find the generated certificate autorithy in the `.http-mitm-proxy/certs/ca.pem` dir, relative to the current working directory.
To trust it on mac OS X, you have to open it in Keychains manager and trust it as an autorithy.

**Comment by shaunlebron** [comment]
Thank you for walking me through it!  I was able to get Chrome to use the proxy through your steps, but the code in the example seems to fail to process the google search page:

<img width="1246" alt="screen shot 2016-05-06 at 11 43 13 am" src="https://cloud.githubusercontent.com/assets/116838/15079703/c34906cc-137f-11e6-8d38-bfe559a12afc.png">

The text being replaced by the example code seems to result in some non-printable characters, so that seems suspect.  To test this further, I modified [examples/processFullResponseBody.js:25] with the following:

[examples/processFullResponseBody.js:25]:https://github.com/joeferner/node-http-mitm-proxy/blob/cc1ea9a81e322fe5cd201ddc822ba2eedcec63d0/examples/processFullResponseBody.js#L25

```js
body = body.toString().replace('</body>','<script>console.info("Hello World");</script></body>');
```

and it works for <https://news.ycombinator.com/> for example, but still fails on the google search page as before.  So it seems particular to however this page is being rendered.

**Comment by felicienfrancois** [comment]
This may be because of the certificate autorithy which is not trusted.
Did you either disabled web security or trusted the certificate autority like I explained ?

**Comment by felicienfrancois** [comment]
Also do you have error logs in the proxy ?

**Comment by shaunlebron** [comment]
I trusted the certificate authority like you explained.  Otherwise, it wouldn't just be the google's search page that would fail.  Everything else works.

Error log after typing "foo" into Chrome's URL/search bar:

```
$ node examples/modifyGoogle.js
listening on 8081
PROXY_TO_SERVER_REQUEST_ERROR on /: { [Error: getaddrinfo ENOTFOUND foo foo:80]
  code: 'ENOTFOUND',
  errno: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'foo',
  host: 'foo',
  port: 80 }
starting server for www.google.com
https server started for www.google.com on 64782
```

**Comment by felicienfrancois** [comment]
The error log is normal as Google Chrome try first to reach the host "foo" which obviously don't work (error ENOTFOUND) and then search on google.

Regarding the initial error, do you get something readable when printing `body.toString()` to console?

**Comment by felicienfrancois** [comment]
Do you still get this error on google?
I'm not able to reproduce but it may be due to the gunzip module.

The `ERR_CONTENT_DECODING_FAILED` error usually happens when the HTTP request headers claim that the content is gzip encoded, but it isn't.

The gunzip module decode gzip encoded responses and remove the `content-encoding` header so the error should not happen.
Can you please look at response headers in the network tab of chrome dev tools?

**Comment by shaunlebron** [comment]
> Regarding the initial error, do you get something readable when printing body.toString() to console?

No, it mucks my console with non-printable characters.

> Do you still get this error on google?

I tried again with latest master. Still getting the same result.

> Can you please look at response headers in the network tab of chrome dev tools?

```
HTTP/1.1 200 OK
content-type: text/html; charset=UTF-8
date: Wed, 11 May 2016 05:29:09 GMT
expires: -1
cache-control: private, max-age=0
content-encoding: sdch, gzip
server: gws
x-xss-protection: 1; mode=block
x-frame-options: SAMEORIGIN
alternate-protocol: 443:quic
alt-svc: quic=":443"; ma=2592000; v="33,32,31,30,29,28,27,26,25"
transfer-encoding: chunked
connection: close
```

Looks like it's using `sdch` encoding in addition to `gzip`?

**Comment by felicienfrancois** [comment]
> Looks like it's using sdch encoding in addition to gzip?

Yes, that's why.

I just made a fix #84.
Can you please checkout the branch [patch-1](https://github.com/joeferner/node-http-mitm-proxy/tree/patch-1) and confirm it fix your issue?

**Comment by shaunlebron** [comment]
I thought it wasn't working until I started placing log statements in the gunzip.onRequest and onResponse functions, then it started working again.  (strange node cache problems?)

did a clean checkout of the repo, and everything is working great.  Thanks @felicienfrancois!

![screen shot 2016-05-11 at 11 14 24 am](https://cloud.githubusercontent.com/assets/116838/15188252/478dc024-176a-11e6-8700-a5aa5f7231f7.png)
